### ISSUE TITLE
Insert size plot question

### ISSUE THREAD

**Issue by ewels** [s-andrews/BamQC#14]
Hello,

I have a question about what the insert size plot is calculating exactly. I ask because I get the same profile, but different numbers, with Picard [CollectInsertSizeMetrics](https://broadinstitute.github.io/picard/command-line-overview.html#CollectInsertSizeMetrics):

**BamQC:**
![image](https://cloud.githubusercontent.com/assets/465550/11240895/84137ca8-8df6-11e5-8c2a-e00a0d4d870d.png)

**Picard:**
![image](https://cloud.githubusercontent.com/assets/465550/11240918/b6476e50-8df6-11e5-95fb-74537b311d0d.png)

These were run on the same input file. Any idea why they're different?

Phil

**Comment by pdp10** [comment]
I don't know as I didn't create this analysis. 
For the y axis, it seems that the tail 0s are removed. The x-axis seems not consistent though.. 
Thanks for pointing this out. 
Piero

**Comment by s-andrews** [comment]
Ha!  That'll be a bug.  I think there's a regex which removes tailing zeros from when all the plots used to be using floating point on the y axis.  Guess it doesn't do the right thing when it's integers!

**Comment by pdp10** [comment]
I had a quick look at this. The class calculating that distribution is "CalculateDistribution" in uk.ac.babraham.BamQC.Utilities . 
I extracted this code from the module InsertLengthDistribution to make it reusable, but I did not write it. 
It seems that the the tail 0s are actually not removed.

**Comment by ewels** [comment]
Ok, nice. I'm not really fussed about the y axis to be honest (I'm not actually sure what those numbers refer to?), I'm more worried about the x axis. BamQC is reporting a median insert size of ~80 bp, whereas Picard comes in at around 220 bp...

**Comment by pdp10** [comment]
The class CalculateDistribution seems to work correctly to me. You can play with the BIN_SIZE parameter to have more or less values for each bar. This changes the y-axis, but certainly not as much as the plot shown using Picard. The x-axis is still the same though. 
The values for x and y axis labels are not altered either. They come from the distribution. In particular:
- y-axis : this is the sum of percent values and depends on BIN_SIZE. This means that if let's say 10 counts are collected (BIN_SIZE=10), each column will be the sum of 10 percentages. 
- x-axis : This is dependent on the values extracted from read.getInferredInsertSize() . I do not see any alteration or re-scaling in the code. 

What do the x and y axes indicate in the Picard plot? There is no axis label.

**Comment by ewels** [comment]
Sorry, I plotted it in excel quickly, my fault. The [Picard documentation](https://broadinstitute.github.io/picard/picard-metric-definitions.html#InsertSizeMetrics) isn't super clear, but the column headers are `insert_size` and `All_Reads.fr_count`.

My understanding is that the insert size is measured from the first bp of read 1 to the first bp of read 2 (everything sequenced, the whole fragment excluding the adapters). For picard, they y axis is then the number of fragments found with that exact length.

Note that there is some confusion about this with other tools - for instance, sometimes 'insert size' is used to describe the distance _in-between_ the two reads (eg. the [RSeQC inner distance plot](http://rseqc.sourceforge.net/#inner-distance-py)) - not as helpful in my mind, as dependent on sequencing length. See seqanswers posts about this [here](http://seqanswers.com/forums/showthread.php?t=15511) and [here](http://seqanswers.com/forums/showthread.php?t=13098).

I wonder if this is what's happening in BamQC? It would explain the lower numbers. But I would expect some overlapping reads in my example above and the axis doesn't go down to 0 / negative numbers. So maybe not.

**Comment by pdp10** [comment]
Thanks for explaining me this. I think we have to go through this more carefully. Surely we need to describe this clearly in the help.

**Comment by ewels** [comment]
No problem! Yes, that would be good :books: :+1:

**Comment by s-andrews** [comment]
We don't calculate that value.  It's taken from the TLEN field in the BAM file, so it's whatever the read mapper puts in there.  It should be the full inferred insert length - at least that's what the docs say.

**Comment by ewels** [comment]
Which docs? ;) Presumably this could differ for various aligners? These plots were done with Bismark anyway, so presumably bowtie2. Any ideas why this reported length would be different to Picard? I'll see if I can open a downsampled version in SeqMonk to plot the same thing for extra validation.

**Comment by s-andrews** [comment]
It shouldn¹t differ for different aligners - the docs are pretty specific about what it¹s supposed to mean.

I know that we did have problems when we were using this value in seqmonk as there were cases where the value reported was technically correct, but not very useful (can¹t remember the details sorry).

Simon.

**Comment by ewels** [comment]
Ah ok. I've heard of problems before when reads are mapped FF or RR instead of FR - distance is taken from first bp of each, so misses a read in FF or RR orientation.

Anyway.. still no closer to figuring out why there's a difference in the plots?

Haven't managed to get SeqMonk to load a file yet. Would be nice if there was an option when importing BAM files to stop after [x] reads :wink:

**Comment by s-andrews** [comment]
There is.

samtools view [file] | head -100000 > small.sam

..and then import that.

:-)

**Comment by ewels** [comment]
Yup, that's pretty much what I was already running (`-s 0.1` instead of `head`). Yours will be much faster though, I'll kill mine and try again.

**Comment by ewels** [comment]
![image](https://cloud.githubusercontent.com/assets/465550/11247182/3d57894c-8e1b-11e5-9984-cf7c75e1e848.png)

Picard wins! Thankfully, as I wanted this data to have long insert sizes.

**Comment by s-andrews** [comment]
What do you get if you extract the tlen values and plot a histogram though?

**Comment by ewels** [comment]
I need to go home! Do you have a command kicking around to extract the `TLEN` values before I start googling? I'll have a go tonight / tomorrow.

**Comment by s-andrews** [comment]
It¹s a tab delimited file.  Just do it in excel!

**Comment by ewels** [comment]
You mean the BamQC text output? But surely that doesn't tell us anything new? I thought you meant extract the `TLEN` flags from the BAM file and plot them?

Data from `bamqc_data.txt`:
![image](https://cloud.githubusercontent.com/assets/465550/11247786/eaa38e1e-8e1d-11e5-8e93-2c8102c14956.png)

**Comment by s-andrews** [comment]
I did.  SAM files are tab delimited files.

**Comment by ewels** [comment]
Yes, yes they are.. Sorry :)

Ok, histogram of the subset of `TLEN` values. I used `=abs()` as read 2 values are negative (so double counts):

![image](https://cloud.githubusercontent.com/assets/465550/11254810/2343bf62-8e43-11e5-88bf-15c60c7329c7.png)

..so looks like BamQC is doing something funny to these numbers?

**Comment by pdp10** [comment]
Does Picard process only proper paired reads for that analysis? 

In BamQC, only reads having the flags getReadPairedFlag() and getProperPairFlag() set to TRUE are parsed. Maybe this is the difference. 
If you suspect this could be reasonable, could you temporarily comment that test in BamQC and see whether the resulting plot is closer to the Picard plot, please? 

Class:  uk/ac/babraham/BamQC/Modules/InsertLengthDistribution.java
Method: public void processSequence(SAMRecord read)

**Comment by ewels** [comment]
I'd be fairly surprised if this is the cause, as I would imagine that such an effect would cause BamQC to give _longer_ insert lengths instead of shorter. I'm at a meeting now but will try to test your suggestion in a bit..

**Comment by s-andrews** [comment]
It's more likely to be a screw up on our end.  Let me take a look.

**Comment by pdp10** [comment]
Thanks to both of you!

**Comment by pdp10** [comment]
I examined the BamQC code a bit better. 
The x-axis groups the lengths using a parameter BIN_SIZE. I made some test with this parameter set to 1 (meaning that now the x-axis directly reflects the value from read.getInferredInsertSize() in SAMRecord). This configuration is currently committed. 

When you (@ewels) have a change, could you let me know whether the plot is closer to the picard result using the above data, please? 
Thanks
